design/client-initiated-backchannel-authentication-flow.md

```diff
@@ -0,0 +1,494 @@
+
+1. Inside CIBA Protocol  : Backchannel Auth Req/Res
+2. Outside CIBA Protocol : AuthN/AuthZ by AD
+3. Inside CIBA Protocol  : Token Req/Res
+
+2 and 3 might be executed in reverse order. 3 might be executed several times until 2 finishes.
+
+CIBA protocol specification requires AuthN/AuthZ by AD but does not define how to do it, namely it is out of scope of CIBA Protocol specification.
+
+In Backchannel Auth Req/Res part, by following CIBA protocol, keycloak processes a backchannel authentication request from Consumption Device (CD) and returns its response. Beside that, keycloak invokes AuthN/AuthZ by AD part.
+
+In AuthN/AuthZ by AD part, keycloak itself or other entity requests an AD to authenticate and get consents from an end-user and inform Token Req/Res part of its result. Besides that, it executes all the processes that need to be executed after the end-user logged-in like creating User Session and Authenticated Client Session that are needed to generate tokens in Token Req/Res part.
+
+In Token Req/Res part, by following CIBA protocol, keycloak processes a token request from CD and returns its response based on the information received from AuthN/AuthZ by AD part.
+
+### Inside CIBA Protocol
+
+This section describes which features defined by CIBA protocol specification are supported or not by this version of the design document.
+
+#### Backchannel Authentication Endpoint
+
+Keycloak's Backchannel Authentication Endpoint basically follows [7. Backchannel Authentication Endpoint](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#rfc.section.7) and [13. Authentication Error Response](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#rfc.section.13) .
+
+URI of this endpoint is as follows :
+
+`{scheme}://{host}[:{port}]/auth/realms/{realm}/protocol/openid-connect/backchannelAuthn`
+
+At this endpoint, the following security tasks are done the same as at Token Endpoint :
+
+* CORS
+* Check TLS
+* Check Realm
+* Check Client (Client Authentication)
+* Check User (is existed, is activated)
+* Prohibit Caching on HTTP Client
+  - Cache-Control : no-store
+  - Pragma : no-cache
+
+#### Token Endpoint
+
+Keycloak's existing Token Endpoint basically follows [10.1. Token Request Using CIBA Grant Type](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#rfc.section.10.1) and [11. Token Error Response](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#rfc.section.11) .
+
+#### Conveying Information on End-User to be Authenticated from CD
+
+On CIBA Protocol specification Section [7.1. Authentication Request](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#auth_request) :
+
+  - Request Parameters
+    * login_hint       : supported
+    * login_hint_token : NOT supported
+    * id_token_hint    : NOT supported
```

Comment:
To identify the user to be authenticated by AD, it is sufficient that at least one of these options be supported. I thought `login_hint` is the simplest option so that I decided to support only `login_hint` option.

But re-considering this matter, it is not so much difficult to support `id_token_hint` because the current keycloak has already supported `id_token_hint` defined in OIDC.

Therefore, I'd like to support `login_hint` and `id_login_hint` . WDYT?

Reply:
+1

Reply:
@mposolda Could you please refer to the review comments for "AuthN/AuthZ User Resolver" about `login_hint`, `login_hint_token` and `id_token_hint` support?